### PR TITLE
Update build process - make Wayland build optional

### PR DESCRIPTION
**Pull request by elParaguayo**
This is an attempt to fix some issues that users are having with building qtile and not wanting the Wayland backend.

The PR removes `pywlroots` from the build dependencies and uses a custom build backend to inject the requirement if the wayland backend is specifically requested.

The builder will also raise an error if the backend is requested but `pywroots` is not installed. This is useful for `--no-build-isolation` flag as that builds against pre-installed libraries.

Comments welcome.

### CODE REVIEW

File: builder.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+from setuptools import build_meta as _orig
+from setuptools.build_meta import *  # noqa: F401,F403
+
+WAYLAND_DEPENDENCIES = ["pywlroots>=0.16.4,<0.17.0"]
+
+
+def wants_wayland(config_settings):
+    if config_settings:
+        for key in ["Backend", "backend"]:
+            if config_settings.get(key, "") in ("Wayland", "wayland"):
```
Comment (jwijenbergh) [inline builder.py:31]: config_settings.get(key, "").lower() == "wayland"

Comment (jwijenbergh) [reply builder.py:31]: same for backend maybe?

Comment (jwijenbergh) [reply builder.py:31]: or maybe we should just make the key --wayland?

Comment (elParaguayo) [reply builder.py:31]: Not possible. There are different builders (`pip`, `python -m build` etc) so the only standard way to pass arguments is the config-settings as that's part of the spec.

Comment (elParaguayo) [reply builder.py:31]: > or: config_settings.get(key, "").lower() == "wayland"

Ha. I thought about this and went with the other option.

Comment (jwijenbergh) [reply builder.py:31]: > Not possible. There are different builders (`pip`, `python -m build` etc) so the only standard way to pass arguments is the config-settings as that's part of the spec.

ah yeah I meant --config-setting wayland or --config-setting --wayland (but that would probably not work). Let's leave it like this and maybe do the lower thing but that's optional

**Comment by jwijenbergh** [comment]
if only there was a neater way to have optional build requires. But this seems like the way to go https://pip.pypa.io/en/stable/reference/build-system/pyproject-toml/#build-time-dependencies

if CI passes LGTM

**Review by jwijenbergh — APPROVED** [review]

**Comment by cr1ogen** [comment]
If you set the installation of wayland as optional, how should Qtile be compiled or installed for those who use wayland as a backend?

**Comment by elParaguayo** [comment]
> If you set the installation of wayland as optional, how should Qtile be compiled or installed for those who use wayland as a backend?

Good question. This is my main concern with this PR as it complicates things a little bit.

Ideally, you should have all the dependencies installed already. In that case, passing the `--no-build-isolation` flag should be enough to let the wayland backend be compiled against your current libraries.

If you don't do that, you can pass `--config-option backend=wayland` which will inject `pywlroots` into the dependencies and it will trigger a failure if it can't build the backend.

**Comment by cr1ogen** [comment]
Couldn't you make an installation script that asks if you want to install the wayland backend?

**Comment by elParaguayo** [comment]
Short answer: no, I don't think so.

What's your main concern? To have Wayland working, you need all dependencies installed beyond just python packages. This change doesn't change anything to do with that but it does remove an unnecessary dependency being installed for people using the x11 backend. 

We can list what packages need to be installed on a user's system but I don't think we should be providing any form of installation script.

**Comment by jwijenbergh** [comment]
I don't really see the problem here, it's just a single flag and would only apply when building from source. Easy to fix in packaging

**Comment by cr1ogen** [comment]
> Short answer: no, I don't think so.
> 
> What's your main concern? To have Wayland working, you need all dependencies installed beyond just python packages. This change doesn't change anything to do with that but it does remove an unnecessary dependency being installed for people using the x11 backend. 
> 
> We can list what packages need to be installed on a user's system but I don't think we should be providing any form of installation script.

I don't have any problem because I know how to do it, but it would be good for the less experienced, at least who wanted to be a window manager only for developers

**Comment by jwijenbergh** [comment]
> > Short answer: no, I don't think so.
> > What's your main concern? To have Wayland working, you need all dependencies installed beyond just python packages. This change doesn't change anything to do with that but it does remove an unnecessary dependency being installed for people using the x11 backend.
> > We can list what packages need to be installed on a user's system but I don't think we should be providing any form of installation script.
> 
> I don't have any problem because I know how to do it, but it would be good for the less experienced, at least who wanted to be a window manager only for developers

Maybe an install script makes sense that just follows the steps that we document.

**Comment by elParaguayo** [comment]
How do you cater for all the different distros/package managers?

**Comment by jwijenbergh** [comment]
> How do you cater for all the different distros/package managers

I mean a script that runs the necessary pip commands, nothing more

**Comment by cr1ogen** [comment]
> > How do you cater for all the different distros/package managers
> 
> I mean a script that runs the necessary pip commands, nothing more

That's what I meant, sorry if I couldn't explain it.

**Review by tych0 — APPROVED** [review]

**Comment by tych0** [comment]
Any reason not to land this?

**Pull request merged**